### ISSUE TITLE
A question about Kvantum user-customized config file ?

### ISSUE THREAD

**Issue by trmdi** [tsujan/Kvantum#275]
When users modify a Kvantum theme from Kvantum Manager, it will copy that config from `/usr/share/Kvantum/Theme/theme.conf` -> `~/.config/Kvantum/Theme#/theme.conf` (*) ?

This way, when you update a theme's config file, users still use the old version in (*).
Everytime I update Kvantum, I have to reset the theme in KV Manager and modify again some values.

Why don't you keep **only the modified** keys & values in (*) insteed ? (similar to systemd's way)

![image](https://user-images.githubusercontent.com/11857205/46962297-f9580080-d0cc-11e8-8010-d16633692bc5.png)

**Comment by tsujan** [comment]
> Why don't you keep only the modified keys & values in (*) insteed ?

The user may also change key/values that aren't changed by Kvantum Manager; therefore, the customized theme (with "#") should be self-sufficient. Kvantum Manager only changes the "safe" keys but the user could do heavy manual editing.

**Comment by tsujan** [comment]
Try it (= delete the keys you don't change) and see firsthand what will happen!

According to Kvantum's inheritance rules, the other keys will get their values from the default Kvantum theme. Breaking those rules would mean breaking Kvantum itself. And *those rules are very useful*. Without them, theme makers should set the values of all possible key in their kvconfig files and draw all possible elements in their SVG files.

There is a huge difference from systemd and similar settings.

**Comment by tsujan** [comment]
> (just an idea, not the most optimized way)

You idea is completely correct and the function `setParent()` exists for it in `ThemeConfig.h`. However, there are 2 drawbacks:

(1) The code would become exceptionally complex. The function `Style::setTheme()` in `Kvantum.cpp` has already more than 300 lines. `KvantumManager.cpp` would get unmaintainable too.

(2) There are serious concerns about CPU usage.

So, `setParent()` is used only for making the default theme the parent of all themes.

**Comment by trmdi** [comment]
> there are 2 drawbacks:

Hmm... this is a technical problem. I have no more idea.

**Comment by tsujan** [comment]
> this is a technical problem

Not only technical, for developers, but also practical for the end user because of the CPU usage.

**Comment by trmdi** [comment]
Think a bit more about the CPU usage, in the step "use the current theme customized values", I guess there is not so many things to do, because the number of "safe config" provided by Kv Manager is not too many. Right ?

**Comment by tsujan** [comment]
> because the number of "safe config" provided by Kv Manager are not too big

Therefore, the number of the other keys is big and they are the ones that should be checked.

Moreover, if the inheritance of the kvconfig file is extended, that of the SVG file should be too. If an element isn't found in an SVG file, its parent should be checked and if it isn't found in its parent, the parent of its parent should be checked, and so on. These computations take CPU time.

**Comment by trmdi** [comment]
>  that of the SVG file should be too
No, KV Manager is only for modifying kvconfig.

> the parent of its parent should be checked, and so on.

The maximum parents = 2.

**Comment by tsujan** [comment]
> The maximum parents = 2.

In this special case, yes, but it still searches in 2 more files, one with ~500 lines, the other with thousands of lines.

**Comment by tsujan** [comment]
> No, KV Manager is only for modifying kvconfig.

An extra inheritance for the SVG file would be logically imperative. We can't limit ourself to special use cases.

**Comment by trmdi** [comment]
Hmm. New idea.
You would keep the customized values in a separated file.
When kvantum selects the theme, the only additional task is that it would create a new temporary kvconfig in memory that join the original one and the value in the separated file. Other functions will not change.

**Comment by trmdi** [comment]
For example, KvMojave has 500 lines, you need to find a quickesr way to create a temp kvconfig then replace values from the separated customized file, about <100 line -> we have a tmp kvconfig.

But I don't know how much time it takes to do that.

**Comment by tsujan** [comment]
No! Because every time a GUI app is launched, Kvantum should do the extra task of applying changes.

As I implied above, if extended parenting is implemented, limiting it to 2 or 3 parents or special use cases will mean a bad and unmaintainable code. On the other hand, if it isn't limited, it'll mean the possibility of extra CPU usage. Frankly, to jeopardize the current stable state of Kvantum by adding hundreds of lines to its code, I need a much stronger reason.

I close this but will think about a "Save As".

**Comment by trmdi** [comment]
Ok, I think I did make it too complicated. I should think it more simply. :dagger: 
Thanks a lot for your reply.

**Comment by tsujan** [comment]
@trmdi Closing this issue doesn't mean I don't understand your use case. Actually, I had to restore the default settings so many times that, recently, I made Kvantum Manager to revert to the root theme on deleting the customized one (previously, it reverted to Kvantum's default theme).

In short, you are right; this is a small annoyance. The only reasonable way I've found to "fix" it is adding a "Save As". This discussion reminded me of it. So, it was positive :)
